### PR TITLE
Support nargs=-1 for option (cleaner)

### PR DESCRIPTION
**Pull request by tristanwietsma**
#222

### CODE REVIEW

**Comment by untitaker** [comment]
What is the behavior when the user doesn't provide a ``--``? When does it fail?

File: click/parser.py [hunk 1]
```diff
@@ -290,6 +298,13 @@ def _match_long_opt(self, opt, explicit_value, state):
 
         option.process(value, state)
 
+    def _next_flag(self, state):
+        is_flag = lambda v: (v.startswith('-') and len(v) == 2) or v.startswith('--')
```
Comment (tristanwietsma) [inline click/parser.py]: @untitaker relevant to your question

**Comment by tristanwietsma** [comment]
if the user doesn't supply '--' or another flag, then there isn't any way to determine where the variadic portion ends.

this usage of '--' is demonstrated in the source elsewhere, so i'm assuming this pattern is fair game.

i'll agree that it feels somewhat awkward in that it requires extra knowledge about the cli. the alternative to variadic options is a comma-separated string of values to be parsed in the command, which isn't very pretty either. bit of a conundrum.

**Comment by untitaker** [comment]
I just read up on click's documentation. I wonder why anybody would want to use such a form if ``multiple=True`` is already available? The resulting commands also seem a lot more explicit.

**Comment by tristanwietsma** [comment]
the particular use case i wrote it for involved a chain of commands in a bash. output from one script was a list of files with unknown length. using this approach, i was able to set nargs=-1 and just pass that output with a " --" appended. alternatively, i could have preprocessed the output in bash.

i think this boils down to a question of where to put the abstraction. i'm not interested in complicating a great library, so i see where you're going. at the same time, the behavior is only in play with "nargs=-1", so it isn't **that** unreasonable to expect the author of the script using this feature to pass additional information in the usage string.

**Comment by untitaker** [comment]
Hmm, I think it might get extremely funky if a misunderstanding user *passes both multiple and nargs=-1*. I'll just leave this open for now.

**Comment by tristanwietsma** [comment]
if you're strongly against it, please feel free to close. this PR is from 11/24; the feature was useful for me and while I see some users have inquired about it, i'm more than happy to just work of my fork. regards.

**Comment by untitaker** [comment]
I'm not strongly against it, I honestly have no strong opinion at all.

**Comment by warsamebashir** [comment]
+1 please merge this

**Comment by mitsuhiko** [comment]
This is opening a can of worms.  It was never intended for nargs < 0 to even remotely work for options and I am not going to support that.  This is however something that might change in the future if sufficient demand comes up with proper justifications about why that's a good idea.

From my perspective it's a terrible idea because it's impossible to unambiguously parse.  This is what arguments are for.

**Comment by MathieuDuponchelle** [comment]
Re "proper justifications about why that's a good idea."

Consider the following case:

I have a CLI that needs to handle two arbitrarily-long lists of distinct filenames (Let's say a bunch of C source files and a bunch of markdown files)

afaiu, the only two solutions are to either:
+ let the user do my_cli --source source.c --source source.h --markdown-file my_file.markdown
+ let the user do my_cli --sources "source.c,source.h" --markdown-files "my_file.markdown"

The first one seems pretty clumsy to me, and the second one error-prone (obviously things will break if a filename contains a comma)

I would like to be able to do my_cli --sources *.c *h --markdown_files *.markdown , which seems to me like a pretty common usecase.

I'm not sure I do understand the issue with allowing this, afaict it should only be a matter of stopping consumption once the parsing code encounters a recognized "--" option ?

**Comment by untitaker** [comment]
>I would like to be able to do my_cli --sources *.c *h --markdown_files *.markdown , which seems to me like a pretty common usecase.

Why not make all of those arguments, like `mv` and `cp` do?

**Comment by untitaker** [comment]
Consider the following example:

    @click.command()
    @click.option('--lol')
    @click.argument('foo', nargs=-1)
    def foo(lol, foo):
        print(lol)
        print(foo)

If I were to add `nargs=-1` to the option, and execute this:

    foo --lol 1 2 3 4 5 6 7 8 9

Click would have no idea what belongs to the option and what to the argument.

**Comment by MathieuDuponchelle** [comment]
> Why not make all of those arguments, like mv and cp do?

I don't understand that question, I have two separate types of files, I can only use nargs=-1 for the final argument so this doesn't solve my problem.

**Comment by untitaker** [comment]
I see, I assumed that you can only specify one markdown file. What if you specify a single `nargs=-1` argument and detect the filetype by file extension?

**Comment by MathieuDuponchelle** [comment]
> Click would have no idea what belongs to the option and what to the argument.

I'm just stating that there is a pretty clear use case here, even more so when taking wildcards into account, I don't know anything about click's implementation but I do know that argparse handles that correctly.

This issue prevents me from using click at all, even though I'd like to use it because I need to resort to funny hacks in argparse to get what click calls "chained commands", it's a pretty silly reason for not using such a nice library :)

**Comment by MathieuDuponchelle** [comment]
> What if you specify a single nargs=-1 argument and detect the filetype by file extension?

This isn't a heuristic I'd be satisfied with, and doesn't solve the general case anyway

**Comment by untitaker** [comment]
As shown, the general case would be ambiguous. Options with nargs=-1 would only be possible if no arguments are specified, which *is* a can of worms.

**Comment by MathieuDuponchelle** [comment]
I would argue that 

+ This could very well be detected at runtime ("Oh no you can't do that you silly")
+ Or afaiu the initial patch set, a separator would seem pretty fair game

**Comment by MathieuDuponchelle** [comment]
Preventing the usage of both nargs=-1 arguments and options doesn't strike me as particularly bad, as it's obvious this can't be done (without a separator), can be easily detected, and simply extends the functionality to let use cases such as mine be implemented with click

**Comment by tristanwietsma** [comment]
Been a while since I looked at this PR, but I believe "--" signals the
option args are complete. See test cases.

**Comment by MathieuDuponchelle** [comment]
Ack

**Comment by MathieuDuponchelle** [comment]
Anyway @mitsuhiko  to summarize my reason for wanting this is "please let me wildcard more than one parameter" :)

**Pull request closed without merging**